### PR TITLE
Test building with external desul

### PR DESCRIPTION
**Pull request by masterleinad**
Follow-up to #5021, also adds a `KOKKOS_ENABLE_DESUL_ATOMICS_FORCE_EXTERNAL` option.

### CODE REVIEW

File: cmake/kokkos_tpls.cmake [hunk 1]
```diff
@@ -85,6 +85,17 @@ ENDIF()
...
 KOKKOS_IMPORT_TPL(LIBQUADMATH)
 
+IF (KOKKOS_ENABLE_IMPL_DESUL_ATOMICS)
+  IF(KOKKOS_ENABLE_DESUL_ATOMICS_FORCE_EXTERNAL)
+    find_package(desul_atomics REQUIRED QUIET)
+  ELSE()
+    find_package(desul_atomics QUIET)
+  ENDIF()
+  IF(desul_atomics_FOUND)
+    KOKKOS_EXPORT_CMAKE_TPL(desul_atomics)
```
Comment (masterleinad) [inline cmake/kokkos_tpls.cmake:95]: It turned out that we don't properly export the `desul` dependency when using an external one.

File: cmake/kokkos_enable_options.cmake [hunk 2]
```diff
@@ -61,6 +61,7 @@ KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_KERNEL_INSTANTIATIONS OFF "Whether multiple ke
 
 # This option will go away eventually, but allows fallback to old implementation when needed.
 KOKKOS_ENABLE_OPTION(IMPL_DESUL_ATOMICS   ON  "Whether to use desul based atomics - option only during beta")
+KOKKOS_ENABLE_OPTION(DESUL_ATOMICS_FORCE_EXTERNAL OFF "Whether to enfore finding an external desul installation")
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:64]: I dislike the option name.
I just came across https://cmake.org/cmake/help/latest/variable/CMAKE_REQUIRE_FIND_PACKAGE_PackageName.html
Wouldn't you agree this would make the new option added here obsolete?

Comment (masterleinad) [reply cmake/kokkos_enable_options.cmake:64]: Yes, if we can rely on CMake >=3.22 the new option makes this obsolete.
I mainly introduced it here for unit testing, though, and CMake is not that new there.

Comment (PhilMiller) [inline cmake/kokkos_enable_options.cmake:64]:
```suggestion
KOKKOS_ENABLE_OPTION(DESUL_ATOMICS_FORCE_EXTERNAL OFF "Whether to enforce finding an external desul installation")
```

Comment (masterleinad) [reply cmake/kokkos_enable_options.cmake]: The option works fine for me using `CMake 3.23.2` but I agree with @nmm0 that it makes sense to force the user to tell us to use an external or internal package (with the default being internal).

File: cmake/kokkos_tpls.cmake [hunk 3]
```diff
@@ -85,6 +85,11 @@ ENDIF()
 KOKKOS_IMPORT_TPL(ROCM INTERFACE)
 KOKKOS_IMPORT_TPL(LIBQUADMATH)
 
+IF (KOKKOS_ENABLE_IMPL_DESUL_ATOMICS AND KOKKOS_ENABLE_DESUL_ATOMICS_EXTERNAL)
```
Comment (dalg24) [inline cmake/kokkos_tpls.cmake:88]:
```suggestion
IF(Kokkos_ENABLE_IMPL_DESUL_ATOMICS AND Kokkos_ENABLE_DESUL_ATOMICS_EXTERNAL)
```
You have not addressed my previous comment about the case of these variables https://github.com/kokkos/kokkos/pull/5053#discussion_r886123158

File: .github/workflows/continuous-integration-workflow.yml [hunk 4]
```diff
@@ -50,11 +65,13 @@ jobs:
       - name: maybe_use_external_gtest
         if: ${{ matrix.distro == 'ubuntu:latest' }}
         run: sudo apt-get update && sudo apt-get install -y libgtest-dev
-      - name: CMake
+      - name: Configure Kokkos
         run: |
           cmake -B builddir \
             -DCMAKE_INSTALL_PREFIX=/usr \
+            -Ddesul_atomics_ROOT=/usr/desul-install/lib/cmake/desul/ \
```
Comment (dalg24) [inline .github/workflows/continuous-integration-workflow.yml:72]:
```suggestion
            -Ddesul_atomics_ROOT=/usr/desul-install/ \
```

File: .github/workflows/continuous-integration-workflow.yml [hunk 5]
```diff
@@ -37,6 +37,21 @@ jobs:
       # see https://github.com/actions/virtual-environments/issues/3812
       options: --security-opt seccomp=unconfined
     steps:
+      - name: Checkout desul
```
Comment (cz4rs) [inline .github/workflows/continuous-integration-workflow.yml:40]: If we want to test both external and internal, this would need to be protected with something similar to `maybe_disable_death_tests` or `maybe_use_external_gtest` steps below.

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:40]: The remainder of our CI, especially everything using `jenkins`, uses the internal one. I would prefer not to add another option but I could do it if people prefer.

File: .github/workflows/continuous-integration-workflow.yml (hunk 4 above)
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:72]: That didn't work. I guess `CMake` expected the path to look like "/usr/desul-install/lib/cmake/desul_atomics".

File: cmake/kokkos_tpls.cmake [hunk 6]
```diff
@@ -85,6 +85,17 @@ ENDIF()
 KOKKOS_IMPORT_TPL(ROCM INTERFACE)
 KOKKOS_IMPORT_TPL(LIBQUADMATH)
 
+IF (KOKKOS_ENABLE_IMPL_DESUL_ATOMICS)
+  IF(KOKKOS_ENABLE_DESUL_ATOMICS_FORCE_EXTERNAL)
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply cmake/kokkos_tpls.cmake]: I am seeing
>$ git grep "Kokkos_ENABLE" cmake | wc -l
37
$ git grep "KOKKOS_ENABLE" cmake | wc -l
165

I think we only introduce the all-uppercase version for the `#cmakedefine`s. And otherwise it's safer to use the `Kokkos_*` version in `CMake` because that works even if the Kokkos option hasn't been called yet (since it's a `CACHE` variable).
Let me know if you want me to change. For this one, I mostly only went with what was already there.

[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/kokkos_tpls.cmake]: Yes change it

File: .github/workflows/continuous-integration-workflow.yml (hunk 4 above)
Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml:97]: Yes we need to fix that on the diesel side...

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:97]: I would prefer addressing that in a follow-up [pull request, possibly including a pull request to desul.

File: cmake/kokkos_tpls.cmake (hunk 3 above)
Comment (masterleinad) [reply cmake/kokkos_tpls.cmake]: We are using `Kokkos_ENABLE_IMPL_DESUL*` etc. now.

File: .github/workflows/continuous-integration-workflow.yml (hunk 4 above)
Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml:97]: Yes sorry I did not mean to imply that is a blocker for this PR

File: cmake/kokkos_tpls.cmake [hunk 7]
```diff
@@ -85,6 +85,11 @@ ENDIF()
 KOKKOS_IMPORT_TPL(ROCM INTERFACE)
 KOKKOS_IMPORT_TPL(LIBQUADMATH)
 
+IF (Kokkos_ENABLE_IMPL_DESUL_ATOMICS AND Kokkos_ENABLE_DESUL_ATOMICS_EXTERNAL)
+  find_package(desul_atomics REQUIRED QUIET)
+  KOKKOS_EXPORT_CMAKE_TPL(desul_atomics)
+ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_tpls.cmake:91]: On second thought, we do not need to introduce an option for this since there is native support for it
* [`CMAKE_DISABLE_FIND_PACKAGE_PackageName`](https://cmake.org/cmake/help/v3.16/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html#cmake-disable-find-package-packagename) to disable the `find_package()` call
* [`CMAKE_REQUIRE_FIND_PACKAGE_PackageName`](https://cmake.org/cmake/help/v3.23/variable/CMAKE_REQUIRE_FIND_PACKAGE_PackageName.html#cmake-require-find-package-packagename) (since CMake 3.22) to turn it into a required dependency

Comment (masterleinad) [reply cmake/kokkos_tpls.cmake:91]: It's just that I can't have the behavior I want for the CI (force finding an external library) with its `CMake` version. Since this is also what users might want, I would prefer the additional CMake variable that can be deprecated/removed once we require CMake 3.22.

Of course, I'm happy to go with the majority vote.

Comment (JBludau) [reply cmake/kokkos_tpls.cmake:91]: I'll try to add my different views on the issue, as a system admin, as a normal user and as a developer.

system admin:
If I install libraries on a system and this installation is possibly shared (e.g. via a module system) to a cluster or just to multiple computers that are used by "common computer knowledge users", I am basically trying to take the configuration and compiling off their desks and onto my desk to help them focus on their work. In this scenario I definitely want the configuration to fail (if the output of cmake is longer than one page) or at least emit a highlighted warning as soon as I would have anything I am dependent on in order to decide if I am okay with an internal or want to externalize it. That said, if I have the option to use a package manager I can offload the hassle of thinking about this to the manager knowing it takes care of this internally (if it can).

normal user who can not offload to the sysadmin:
An error at configure time due to failing to find a lib would cause irritation and frustration that I need to take care of this. If it would be just a warning that an internal version is used i would not care but would be blind to the consequences. Even more so if it configures with just a status message. If it would give me an error but also hint to the possibility of using an internal version and tell me the configure options, I would be happy again.

kokkos developer/advanced user:
As long as I am aware of what it means when I am using the internal version I am fine with it. So having no warnings in the cmake output once I specify which option to use would be fine. For a CI I would like to get an error if something warns but this could be handled with `-Werror=dev` of cmake.

tl:dr
I would like to see an error if find_package fails. The error message should give the argument to use the internal desul for normal users that don't care together with a short explanation. If the internal is used I would still print a highlighted warning in the configure output. For the external I am fine with having no warning but just a status, as this is the "expected" behavior. I am in favor of not adding another variable and stick to the cmake commands.

File: .github/workflows/continuous-integration-workflow.yml (hunk 4 above)
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:97]: See https://github.com/desul/desul/pull/65.

File: cmake/kokkos_tpls.cmake (hunk 7 above)
Comment (masterleinad) [reply cmake/kokkos_tpls.cmake:91]: We decided to add a `CMake` variable that forces using an external installation. I suggested `Kokkos_ENABLE_DESUL_ATOMICS_EXTERNAL` here but don't care too much if someone has a better suggestion.

File: cmake/kokkos_tpls.cmake [hunk 8]
```diff
@@ -85,6 +85,11 @@ ENDIF()
 KOKKOS_IMPORT_TPL(ROCM INTERFACE)
 KOKKOS_IMPORT_TPL(LIBQUADMATH)
 
+IF (Kokkos_ENABLE_IMPL_DESUL_ATOMICS AND Kokkos_ENABLE_DESUL_ATOMICS_EXTERNAL)
+  find_package(desul REQUIRED QUIET COMPONENTS atomics)
```
Comment (dalg24) [inline cmake/kokkos_tpls.cmake:89]: Why QUIET?

**Review by dalg24 — COMMENTED** [review]
looks good other than the `QUIET` comment

File: cmake/kokkos_tpls.cmake (hunk 8 above)
Comment (masterleinad) [reply cmake/kokkos_tpls.cmake:89]: I just kept what we had before. `REQUIRED QUIET` still prints the usual error when the package is not found but doesn't print anything if the package is found.

Comment (masterleinad) [reply cmake/kokkos_tpls.cmake]: You get 
```
CMake Error at cmake/kokkos_tpls.cmake:89 (find_package):
  By not providing "Finddesul_atomics.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "desul_atomics", but CMake did not find one.

  Could not find a package configuration file provided by "desul_atomics"
  with any of the following names:

    desul_atomicsConfig.cmake
    desul_atomics-config.cmake

  Add the installation prefix of "desul_atomics" to CMAKE_PREFIX_PATH or set
  "desul_atomics_DIR" to a directory containing one of the above files.  If
  "desul_atomics" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:244 (INCLUDE)
  CMakeLists.txt:188 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```
if `desul` wasn't found.

I don't have a strong opinion on using `QUIET` or not but I think it makes sense conceptionally since we print anyway if a bundled or external package is used.
(At the moment, we don't print anything anyway without `QUIET`.)

**Comment by masterleinad** [comment]
I cleaned the history.

File: cmake/kokkos_tpls.cmake (hunk 8 above)
Comment (dalg24) [reply cmake/kokkos_tpls.cmake]: Drop it please

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Review by JBludau — APPROVED** [review]
Looks good but we should revise this when we bump up to cmake 3.22.

**Review by cz4rs — APPROVED** [review]
Looks good to me 👍

**Comment by dalg24** [comment]
Doesn't compile

**Comment by cz4rs** [comment]
> Doesn't compile

I assume that the last commit should be removed:

> [Daniel Arndt]  [7:56 PM]
Don't mind the last commit. This can be done in a follow-up.

**Comment by masterleinad** [comment]
> I assume that the last commit should be removed:
> 
> > [Daniel Arndt]  [7:56 PM]
> > Don't mind the last commit. This can be done in a follow-up.

Yes, I removed it now in anticipation of using the `COMPONENTS` in a follow-up pull request.

**Comment by masterleinad** [comment]
> Looks good but we should revise this when we bump up to cmake 3.22.

See https://github.com/kokkos/kokkos/issues/5131.
